dataset/reduction.cpp

```diff
@@ -59,9 +59,35 @@ DataArray mean(const DataArrayConstView &a, const Dim dim) {
       a, [](auto &&... _) { return mean(_...); }, dim, a.masks());
 }
 
+DataArray mean(const DataArrayConstView &a) {
+  return sum(a) / scale_divisor(a.data(), a.masks());
+}
+
 Dataset mean(const DatasetConstView &d, const Dim dim) {
   return apply_to_items(
       d, [](auto &&... _) { return mean(_...); }, dim);
 }
 
+Dataset mean(const DatasetConstView &d) {
+  return apply_to_items(d, [](auto &&... _) { return mean(_...); });
+}
+
+DataArray nanmean(const DataArrayConstView &a, const Dim dim) {
+  return apply_to_data_and_drop_dim(
+      a, [](auto &&... _) { return nanmean(_...); }, dim, a.masks());
```

Comment:
Is this the same as:
```suggestion
  return nansum(a, dim) / sum(!isnan(a), dim);
```
Would it remove the need for the `nanmean` functions in `dataset/variable_reduction.cpp`?

Reply:
`DataArray` support isn't present for these functions `(isnan, isinf)`. I think we should add this as part of a follow up and refactor internals to use it.

Reply:
I had trouble understanding the current logic, so maybe we can add it now? It should be trivial copy and paste, since it is just a unary operation.